### PR TITLE
shell: shutdown dialog schedule date

### PR DESCRIPTION
**Pull request by fridex**
Based on https://github.com/cockpit-project/cockpit/issues/33

### CODE REVIEW

**Comment by stefwalter** [comment]
@andreasn Could you review the implementation of  design.

@fridex Thanks!

**Comment by andreasn** [comment]
Tested it a bit, and works as expected. Looks really nice!

**Comment by andreasn** [comment]
I did realize that we might want some kind of message for when a scheduled reboot or shutdown is about to happen. At least in the shutdown dialog itself. But we can do that as a separate pull request if that's better.

**Comment by stefwalter** [comment]
> I did realize that we might want some kind of message for when a scheduled reboot or shutdown is about to happen. At least in the shutdown dialog itself. But we can do that as a separate pull request if that's better.

Yes, and we need to do this regardless of what mechanism is used to initiate the shutdown. Bonus points for actually mirroring writes to the TTY ... so one would see terminal announcemenst appear in the top bar of cockpit as notifications.

**Comment by andreasn** [comment]
I found an earlier issue about that issue #1153

File: pkg/shell/cockpit-server.js [hunk 1]
```diff
@@ -824,54 +849,120 @@ PageShutdownDialog.prototype = {
...
+                        dfd.resolve(when);
+                    });
+            } else {
+                dfd.resolve(delay);
+            }
+
+            return dfd.promise();
+        }
+
+        if (! this.validate())
```
Comment (stefwalter) [inline pkg/shell/cockpit-server.js]: Extra space.

File: pkg/shell/cockpit-server.js [hunk 2]
```diff
@@ -824,54 +849,120 @@ PageShutdownDialog.prototype = {
...
+            } else {
+                when = "+" + when;
+            }
+
+            var arg = (op == "shutdown") ? "--poweroff" : "--reboot";
+            var message = $("#shutdown-message").val();
+            cockpit.spawn(["shutdown", arg, when, message], { superuser: true })
+                .fail(function(ex) {
+                    $('#shutdown-dialog').modal('hide');
+                    shell.show_unexpected_error(ex);
```
Comment (stefwalter) [inline pkg/shell/cockpit-server.js]: This should be shown in the dialog, as described here: https://github.com/cockpit-project/cockpit/wiki/HIG:-Dialogs

Here's an example of showing a general failure in a dialog. https://github.com/cockpit-project/cockpit/blob/master/pkg/playground/patterns.html#L85

In addition the dialog should not close on failure.

File: pkg/shell/cockpit-server.js [hunk 3]
```diff
@@ -824,54 +849,120 @@ PageShutdownDialog.prototype = {
...
+            if (when < 0) {
+                $('#shutdown-parse-error').text(_("Cannot schedule event in the past"));
+                self.show_errors(true);
+            } else {
+                when = "+" + when;
+            }
+
+            var arg = (op == "shutdown") ? "--poweroff" : "--reboot";
+            var message = $("#shutdown-message").val();
+            cockpit.spawn(["shutdown", arg, when, message], { superuser: true })
```
Comment (stefwalter) [inline pkg/shell/cockpit-server.js]: Please us the dialog 'wait' behavior to show a spinner while executing the "shutdown" command: https://github.com/cockpit-project/cockpit/blob/master/pkg/playground/patterns.html#L119

You can pass a promise as the second argument.

File: pkg/shell/cockpit-server.js [hunk 4]
```diff
@@ -824,54 +849,120 @@ PageShutdownDialog.prototype = {
...
         }
+
+        this.update_minutes();
         this.update();
     },
 
+    show_errors: function(show) {
+        $('#shutdown-parse-error').css('visibility', show ? 'visible' : 'hidden');
+        $('#shutdown-action').prop('disabled', show);
+        $("#shutdown-time").toggleClass("has-error", show);
```
Comment (stefwalter) [inline pkg/shell/cockpit-server.js]: Please use the new dialog behavior functionality here: https://github.com/cockpit-project/cockpit/blob/master/pkg/playground/patterns.html#L93

File: pkg/shell/cockpit-server.js [hunk 5]
```diff
@@ -72,6 +72,13 @@ function ServerTime() {
         }
```
Comment (stefwalter) [inline pkg/shell/cockpit-server.js]: Lets call the 'now' property 'utc_fake_now' and call out its strange behavior in its name.

File: pkg/shell/cockpit-server.js [hunk 6]
```diff
@@ -72,6 +72,13 @@ function ServerTime() {
...
     });
 
+    Object.defineProperty(self, 'raw_now', {
```
Comment (stefwalter) [inline pkg/shell/cockpit-server.js]: And lets call this 'now'. That'll be clearer.

File: pkg/shell/shell.html [hunk 7]
```diff
@@ -2397,14 +2397,26 @@ <h4 class="modal-title"></h4>
...
+                      <div class="col-md-1 text-center nopadding">:</div>
+                      <div class="col-md-5 nopadding">
+                        <input type='text' class="form-control" id="shutdown-time-minutes"/>
+                      </div>
+                    </div>
                   </div>
                 </td>
               </tr>
             </table>
+            <div class="has-error">
```
Comment (stefwalter) [inline pkg/shell/shell.html]: This isn't necessary once you start using the dialog patterns. See above.

File: pkg/shell/cockpit-server.js (hunk 5 above)
Comment (stefwalter) [inline pkg/shell/cockpit-server.js]: Done.

File: pkg/shell/cockpit-server.js (hunk 6 above)
Comment (stefwalter) [inline pkg/shell/cockpit-server.js]: Done.

**Comment by stefwalter** [comment]
Did some work on this in my scheduled-shutdown-date branch. But i'll probably end up rewriting the entire shutdown dialog to use better code.

**Pull request closed without merging**